plugins/woocommerce/src/Internal/Admin/BlockTemplates/AbstractBlockTemplate.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+namespace Automattic\WooCommerce\Internal\Admin\BlockTemplates;
+
+use Automattic\WooCommerce\Admin\BlockTemplates\ContainerInterface;
+use Automattic\WooCommerce\Admin\BlockTemplates\BlockInterface;
+use Automattic\WooCommerce\Admin\BlockTemplates\BlockTemplateInterface;
+
+/**
+ * Block template class.
+ */
+abstract class AbstractBlockTemplate implements BlockTemplateInterface {
+	use BlockContainerTrait;
+
+	/**
+	 * The block cache.
+	 *
+	 * @var BlockInterface[]
+	 */
+	private $block_cache = [];
+
+	/**
+	 * Get a block by ID.
+	 *
+	 * @param string $block_id The block ID.
+	 */
+	public function get_block( string $block_id ): ?BlockInterface {
+		return $this->block_cache[ $block_id ] ?? null;
+	}
+
+	/**
+	 * Caches a block in the template. This is an internal method and should not be called directly
+	 * except for classes that implement BlockContainerInterface, in their add_block() method.
+	 *
+	 * @param BlockInterface $block The block to cache.
+	 *
+	 * @throws \ValueError If a block with the specified ID already exists in the template.
+	 * @throws \ValueError If the block template that the block belongs to is not this template.
+	 *
+	 * @ignore
+	 */
+	public function cache_block( BlockInterface &$block ) {
+		$id = $block->get_id();
+
+		if ( isset( $this->block_cache[ $id ] ) ) {
+			throw new \ValueError( 'A block with the specified ID already exists in the template.' );
+		}
+
+		if ( $block->get_root_template() !== $this ) {
+			throw new \ValueError( 'The block template that the block belongs to must be the same as this template.' );
+		}
+
+		$this->block_cache[ $id ] = $block;
+	}
+
+	/**
+	 * Generate a block ID based on a base.
+	 *
+	 * @param string $id_base The base to use when generating an ID.
+	 * @return string
+	 */
+	public function generate_block_id( string $id_base ): string {
+		$instance_count = 0;
+
+		do {
+			$instance_count++;
+			$block_id = $id_base . '-' . $instance_count;
+		} while ( isset( $this->block_cache[ $block_id ] ) );
```

Comment:
I don't think there's much more context to this.  I did something similar in my initial concept and I think @mattsherman followed the same pattern.

I like that the IDs are somewhat predictable `my-block-1`, `my-block-2`, etc if an ID is not provided and the lookup in the cache itself is `O(1)` while the instance count could be `O(n)` where "n" is the number of the same instances.

A `uid` might not work very well here since these IDs are not persisted and the only way to know this ID would be to parse the entire tree.

```
$template->get_block( `a78b898c-c999-437f-9102-34c85779340d` ) 
```

That said, the predictability of the current block IDs is not a guarantee and could change if the order of insertion changed.

```
$template->get_block( `my-block-1` ) 
```

I'm open for changes on this if you still prefer the UID, though I do think this is something we can iterate on in the short-term if necessary.

Reply:
Iterating on it in the short term sounds good with me, and thanks for the clarification :)